op-dispute-mon/mon/extract/output_agreement_enricher.go

```diff
@@ -30,54 +32,129 @@ type OutputMetrics interface {
+func NewOutputAgreementEnricher(logger log.Logger, metrics OutputMetrics, clients []OutputRollupClient, cl clock.Clock) *OutputAgreementEnricher {
 	return &OutputAgreementEnricher{
 		log:     logger,
 		metrics: metrics,
-		client:  client,
+		clients: clients,
 		clock:   cl,
 	}
 }
 
+type outputResult struct {
+	outputRoot common.Hash
+	isSafe     bool
+	notFound   bool
+	err        error
+}
+
 // Enrich validates the specified root claim against the output at the given block number.
 func (o *OutputAgreementEnricher) Enrich(ctx context.Context, block rpcblock.Block, caller GameCaller, game *monTypes.EnrichedGameData) error {
 	if !game.UsesOutputRoots() {
 		return nil
 	}
-	if o.client == nil {
+	if len(o.clients) == 0 {
 		return fmt.Errorf("%w but required for game type %v", ErrRollupRpcRequired, game.GameType)
 	}
-	output, err := o.client.OutputAtBlock(ctx, game.L2BlockNumber)
-	if err != nil {
-		// string match as the error comes from the remote server so we can't use Errors.Is sadly.
-		if strings.Contains(err.Error(), "not found") {
-			// Output root doesn't exist, so we must disagree with it.
-			game.AgreeWithClaim = false
-			return nil
+
+	results := make([]outputResult, len(o.clients))
+	var wg sync.WaitGroup
+	for i, client := range o.clients {
+		wg.Add(1)
+		go func(i int, client OutputRollupClient) {
+			defer wg.Done()
+			output, err := client.OutputAtBlock(ctx, game.L2BlockNumber)
+			if err != nil {
+				// string match as the error comes from the remote server so we can't use Errors.Is sadly.
+				if strings.Contains(err.Error(), "not found") {
+					results[i] = outputResult{notFound: true}
+					return
+				}
+				results[i] = outputResult{err: err}
+				return
+			}
+
+			outputRoot := common.Hash(output.OutputRoot)
+			results[i] = outputResult{outputRoot: outputRoot}
+
+			// Only check if the output root is safe if it matches the game's root claim
+			if outputRoot == game.RootClaim {
+				safeHead, err := client.SafeHeadAtL1Block(ctx, game.L1HeadNum)
+				if err != nil {
+					o.log.Warn("Unable to verify proposed block was safe", "l1HeadNum", game.L1HeadNum, "l2BlockNum", game.L2BlockNumber, "err", err)
+					// If safe head data isn't available, assume the output root was safe
+					// Avoids making the dispute mon dependent on safe head db being available
+					results[i].isSafe = true
+					return
+				}
+				results[i].isSafe = safeHead.SafeHead.Number >= game.L2BlockNumber
+			}
+		}(i, client)
+	}
+	wg.Wait()
+
+	validResults := make([]outputResult, 0, len(results))
+	syncedResults := make([]outputResult, 0, len(results))
+	for idx, result := range results {
+		if result.err != nil {
+			o.log.Error("Failed to fetch output root", "clientIndex", idx, "l2BlockNum", game.L2BlockNumber, "err", result.err)
+			continue
+		}
+
+		validResults = append(validResults, result)
+
+		if result.notFound {
+			o.log.Warn("Node is out of sync", "clientIndex", idx, "l2BlockNum", game.L2BlockNumber)
```

Comment:
We can't log this yet unfortunately - if the proposal is in the future, all nodes will return not found and we shouldn't log a warning for any of them.

Actually thinking about it some more, there's a corner case here that could lead to unnecessary warnings if we just look at notFound:
1. Current unsafe block is 100
2. Proposal is made for block 105
3. All nodes return notFound so that fine.
4. Time progresses and the unsafe block is now 105.
5. Due to normal replication delays we get a mix of nodes returning notFound and found. So we warn that some are out of sync. But they are actually all agreeing that the proposal is invalid because there was no safe block 105 when the game was created.

We're going to need to rethink how we detect out of sync nodes...